pkg/core/blockchain.go

```diff
@@ -410,6 +413,72 @@ func (bc *Blockchain) init() error {
 	return bc.updateExtensibleWhitelist(bHeight)
 }
 
+// JumpToState is an atomic operation that changes Blockchain state to the one
+// specified by the state sync point p. All the data needed for the jump must be
+// collected by the state sync module.
+func (bc *Blockchain) JumpToState(p uint32) error {
+	bc.lock.Lock()
+	bc.log.Info("jumping to state sync point", zap.Uint32("state sync point", p))
+
+	block, err := bc.dao.GetBlock(bc.headerHashes[p])
+	if err != nil {
+		bc.lock.Unlock()
+		return fmt.Errorf("failed to get current block: %w", err)
+	}
+	err = bc.dao.StoreAsCurrentBlock(block, nil)
+	if err != nil {
+		bc.lock.Unlock()
+		return fmt.Errorf("failed to store current block: %w", err)
+	}
+	bc.topBlock.Store(block)
+	atomic.StoreUint32(&bc.blockHeight, p)
+	atomic.StoreUint32(&bc.persistedHeight, p)
+
+	block, err = bc.dao.GetBlock(bc.headerHashes[p+1])
+	if err != nil {
+		bc.lock.Unlock()
+		return fmt.Errorf("failed to get block to init MPT: %w", err)
+	}
+	if err = bc.stateRoot.JumpToState(&state.MPTRoot{
+		Index: p,
+		Root:  block.PrevStateRoot,
+	}, bc.config.KeepOnlyLatestState); err != nil {
+		return fmt.Errorf("can't perform MPT jump to height %d: %w", p, err)
+	}
+
+	err = bc.contracts.NEO.InitializeCache(bc, bc.dao)
+	if err != nil {
+		bc.lock.Unlock()
+		return fmt.Errorf("can't init cache for NEO native contract: %w", err)
+	}
+	err = bc.contracts.Management.InitializeCache(bc.dao)
+	if err != nil {
+		bc.lock.Unlock()
+		return fmt.Errorf("can't init cache for Management native contract: %w", err)
+	}
+	bc.contracts.Designate.InitializeCache()
+
+	if err := bc.updateExtensibleWhitelist(p); err != nil {
+		bc.lock.Unlock()
+		return err
+	}
+
+	// after current state is updated, we need to remove old state-related data
+	cache := bc.dao.GetWrapped()
+	writeBuf := io.NewBufBinWriter()
+	for index := 0; index < int(p-bc.config.MaxTraceableBlocks); index++ {
+		// ignore error, as some blocks may be missing from storage
+		_ = cache.DeleteBlock(bc.headerHashes[index], writeBuf)
+		writeBuf.Reset()
+	}
+	// TODO: remove outdated NEP17 transfers and NEP17 transfer info, #2096 related.
```

Comment:
How come we have any old data if we're doing the jump only once and even that before the node does any real block processing?

Reply:
I thought that we should also consider the case when the node had been in sync for some time and reached N height. After that, the node was stopped for quite a long time (more than 2*StateSyncInterval). And after that was started again with outdated values in the DB (blocks, transactions, AERs, NEP17 transfers, storage items).

Reply:
We're only doing the jump once. If the node has synchronized to some point already it never jumps again. If it's so stale that it'd better be jumping the operator can always just drop the DB.

Reply:
The genesis block is always processed by the node in a regular manner (if the node starts from an empty DB and trying to sync with other nodes). So we need to remove NEP17-related info for the genesis block. As far as the block itself and AERs. I adjusted this behaviour.